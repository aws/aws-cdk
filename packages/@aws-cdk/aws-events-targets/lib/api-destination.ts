import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import { addToDeadLetterQueueResourcePolicy, bindBaseTargetConfig, singletonEventRole, TargetBaseProps } from './util';

/**
 * Customize the EventBridge Api Destinations Target
 */
export interface ApiDestinationProps extends TargetBaseProps {
  /**
   * The event to send
   *
   * @default - the entire EventBridge event
   */
  readonly event?: events.RuleTargetInput;

  /**
   * The role to assume before invoking the target
   *
   * @default - a new role will be created
   */
  readonly eventRole?: iam.IRole;

  /**
   * Additional headers sent to the API Destination
   *
   * These are merged with headers specified on the Connection, with
   * the headers on the Connection taking precedence.
   *
   * You can only specify secret values on the Connection.
   *
   * @default - none
   */
  readonly headerParameters?: Record<string, string>;

  /**
   * Path parameters to insert in place of path wildcards (`*`).
   *
   * If the API destination has a wilcard in the path, these path parts
   * will be inserted in that place.
   *
   * @default - none
   */
  readonly pathParameterValues?: string[]

  /**
   * Additional query string parameters sent to the API Destination
   *
   * These are merged with headers specified on the Connection, with
   * the headers on the Connection taking precedence.
   *
   * You can only specify secret values on the Connection.
   *
   * @default - none
   */
  readonly queryStringParameters?: Record<string, string>;
}

/**
 * Use an API Destination rule target.
 */
export class ApiDestination implements events.IRuleTarget {
  constructor(
    private readonly apiDestination: events.IApiDestination,
    private readonly props: ApiDestinationProps = {},
  ) { }

  /**
   * Returns a RuleTarget that can be used to trigger API destinations
   * from an EventBridge event.
   */
  public bind(_rule: events.IRule, _id?: string): events.RuleTargetConfig {
    const httpParameters: events.CfnRule.HttpParametersProperty | undefined =
      this.props.headerParameters ??
      this.props.pathParameterValues ??
      this.props.queryStringParameters
        ? {
          headerParameters: this.props.headerParameters,
          pathParameterValues: this.props.pathParameterValues,
          queryStringParameters: this.props.queryStringParameters,
        } : undefined;

    if (this.props?.deadLetterQueue) {
      addToDeadLetterQueueResourcePolicy(_rule, this.props.deadLetterQueue);
    }

    const role = this.props?.eventRole ?? singletonEventRole(this.apiDestination);
    role.addToPrincipalPolicy(new iam.PolicyStatement({
      resources: [this.apiDestination.apiDestinationArn],
      actions: ['events:InvokeApiDestination'],
    }));

    return {
      ...(this.props ? bindBaseTargetConfig(this.props) : {}),
      arn: this.apiDestination.apiDestinationArn,
      role,
      input: this.props.event,
      targetResource: this.apiDestination,
      httpParameters,
    };
  }
}
