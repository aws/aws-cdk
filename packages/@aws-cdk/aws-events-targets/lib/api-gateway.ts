import * as api from '@aws-cdk/aws-apigateway';
import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import { addToDeadLetterQueueResourcePolicy, bindBaseTargetConfig, singletonEventRole, TargetBaseProps } from './util';

/**
 * Customize the API Gateway Event Target
 */
export interface ApiGatewayProps extends TargetBaseProps {

  /**
   * The method for api resource invoked by the rule.
   *
   * @default '*' that treated as ANY
   */
  readonly method?: string;

  /**
   * The api resource invoked by the rule.
   * We can use wildcards('*') to specify the path. In that case,
   * an equal number of real values must be specified for pathParameterValues.
   *
   * @default '/'
   */
  readonly path?: string;

  /**
   * The deploy stage of api gateway invoked by the rule.
   *
   * @default the value of deploymentStage.stageName of target api gateway.
   */
  readonly stage?: string;

  /**
   * The headers to be set when requesting API
   *
   * @default no header parameters
   */
  readonly headerParameters?: { [key: string]: (string) };

  /**
   * The path parameter values to be used to
   * populate to wildcards("*") of requesting api path
   *
   * @default no path parameters
   */
  readonly pathParameterValues?: string[];

  /**
   * The query parameters to be set when requesting API.
   *
   * @default no querystring parameters
   */
  readonly queryStringParameters?: { [key: string]: (string) };

  /**
   * This will be the post request body send to the API.
   *
   * @default the entire EventBridge event
   */
  readonly postBody?: events.RuleTargetInput;

  /**
   * The role to assume before invoking the target
   * (i.e., the pipeline) when the given rule is triggered.
   *
   * @default - a new role will be created
   */
  readonly eventRole?: iam.IRole;
}

/**
 * Use an API Gateway REST APIs as a target for Amazon EventBridge rules.
 */
export class ApiGateway implements events.IRuleTarget {

  constructor(public readonly restApi: api.RestApi, private readonly props?: ApiGatewayProps) {
  }

  /**
   * Returns a RuleTarget that can be used to trigger this API Gateway REST APIs
   * as a result from an EventBridge event.
   *
   * @see https://docs.aws.amazon.com/eventbridge/latest/userguide/resource-based-policies-eventbridge.html#sqs-permissions
   */
  public bind(rule: events.IRule, _id?: string): events.RuleTargetConfig {
    if (this.props?.deadLetterQueue) {
      addToDeadLetterQueueResourcePolicy(rule, this.props.deadLetterQueue);
    }

    const wildcardCountsInPath = this.props?.path?.match( /\*/g )?.length ?? 0;
    if (wildcardCountsInPath !== (this.props?.pathParameterValues || []).length) {
      throw new Error('The number of wildcards in the path does not match the number of path pathParameterValues.');
    }

    const restApiArn = this.restApi.arnForExecuteApi(
      this.props?.method,
      this.props?.path || '/',
      this.props?.stage || this.restApi.deploymentStage.stageName,
    );
    return {
      ...(this.props ? bindBaseTargetConfig(this.props) : {}),
      arn: restApiArn,
      role: this.props?.eventRole || singletonEventRole(this.restApi, [new iam.PolicyStatement({
        resources: [restApiArn],
        actions: [
          'execute-api:Invoke',
          'execute-api:ManageConnections',
        ],
      })]),
      deadLetterConfig: this.props?.deadLetterQueue && { arn: this.props.deadLetterQueue?.queueArn },
      input: this.props?.postBody,
      targetResource: this.restApi,
      httpParameters: {
        headerParameters: this.props?.headerParameters,
        queryStringParameters: this.props?.queryStringParameters,
        pathParameterValues: this.props?.pathParameterValues,
      },
    };
  }

}

