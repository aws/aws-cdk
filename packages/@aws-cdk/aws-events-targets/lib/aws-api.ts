import * as path from 'path';
import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import { metadata } from './sdk-api-metadata.generated';
import { addLambdaPermission } from './util';

/**
 * AWS SDK service metadata.
 */
export type AwsSdkMetadata = {[key: string]: any};

const awsSdkMetadata: AwsSdkMetadata = metadata;

/**
 * Rule target input for an AwsApi target.
 */
export interface AwsApiInput {
  /**
   * The service to call
   *
   * @see https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/index.html
   */
  readonly service: string;

  /**
   * The service action to call
   *
   * @see https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/index.html
   */
  readonly action: string;

  /**
   * The parameters for the service action
   *
   * @see https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/index.html
   *
   * @default - no parameters
   */
  readonly parameters?: any;

  /**
   * The regex pattern to use to catch API errors. The `code` property of the
   * `Error` object will be tested against this pattern. If there is a match an
   * error will not be thrown.
   *
   * @default - do not catch errors
   */
  readonly catchErrorPattern?: string;

  /**
   * API version to use for the service
   *
   * @see https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/locking-api-versions.html
   * @default - use latest available API version
   */
  readonly apiVersion?: string;
}

/**
 * Properties for an AwsApi target.
 */
export interface AwsApiProps extends AwsApiInput {
  /**
   * The IAM policy statement to allow the API call. Use only if
   * resource restriction is needed.
   *
   * @default - extract the permission from the API call
   */
  readonly policyStatement?: iam.PolicyStatement;
}

/**
 * Use an AWS Lambda function that makes API calls as an event rule target.
 */
export class AwsApi implements events.IRuleTarget {
  constructor(private readonly props: AwsApiProps) {}

  /**
   * Returns a RuleTarget that can be used to trigger this AwsApi as a
   * result from an EventBridge event.
   */
  public bind(rule: events.IRule, id?: string): events.RuleTargetConfig {
    const handler = new lambda.SingletonFunction(rule as events.Rule, `${rule.node.id}${id}Handler`, {
      code: lambda.Code.fromAsset(path.join(__dirname, 'aws-api-handler')),
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      uuid: 'b4cf1abd-4e4f-4bc6-9944-1af7ccd9ec37',
      lambdaPurpose: 'AWS',
    });

    if (this.props.policyStatement) {
      handler.addToRolePolicy(this.props.policyStatement);
    } else {
      handler.addToRolePolicy(new iam.PolicyStatement({
        actions: [awsSdkToIamAction(this.props.service, this.props.action)],
        resources: ['*'],
      }));
    }

    // Allow handler to be called from rule
    addLambdaPermission(rule, handler);

    const input: AwsApiInput = {
      service: this.props.service,
      action: this.props.action,
      parameters: this.props.parameters,
      catchErrorPattern: this.props.catchErrorPattern,
      apiVersion: this.props.apiVersion,
    };

    return {
      id: '',
      arn: handler.functionArn,
      input: events.RuleTargetInput.fromObject(input),
      targetResource: handler,
    };
  }
}

/**
 * Transform SDK service/action to IAM action using metadata from aws-sdk module.
 */
function awsSdkToIamAction(service: string, action: string): string {
  const srv = service.toLowerCase();
  const iamService = awsSdkMetadata[srv].prefix || srv;
  const iamAction = action.charAt(0).toUpperCase() + action.slice(1);
  return `${iamService}:${iamAction}`;
}
