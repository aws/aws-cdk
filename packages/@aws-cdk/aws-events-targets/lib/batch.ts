import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import { Names } from '@aws-cdk/core';
import { IConstruct } from 'constructs';
import { addToDeadLetterQueueResourcePolicy, bindBaseTargetConfig, singletonEventRole, TargetBaseProps } from './util';

/**
 * Customize the Batch Job Event Target
 */
export interface BatchJobProps extends TargetBaseProps {
  /**
   * The event to send to the Lambda
   *
   * This will be the payload sent to the Lambda Function.
   *
   * @default the entire EventBridge event
   */
  readonly event?: events.RuleTargetInput;

  /**
   * The size of the array, if this is an array batch job.
   *
   * Valid values are integers between 2 and 10,000.
   *
   * @default no arrayProperties are set
   */
  readonly size?: number;

  /**
   * The number of times to attempt to retry, if the job fails. Valid values are 1â€“10.
   *
   * @default no retryStrategy is set
   */
  readonly attempts?: number;

  /**
   * The name of the submitted job
   *
   * @default - Automatically generated
   */
  readonly jobName?: string;
}

/**
 * Use an AWS Batch Job / Queue as an event rule target.
 * Most likely the code will look something like this:
 * `new BatchJob(jobQueue.jobQueueArn, jobQueue, jobDefinition.jobDefinitionArn, jobDefinition)`
 *
 * In the future this API will be improved to be fully typed
 */
export class BatchJob implements events.IRuleTarget {
  constructor(
    /**
     * The JobQueue arn
     */
    private readonly jobQueueArn: string,

    /**
     * The JobQueue Resource
     */
    private readonly jobQueueScope: IConstruct,

    /**
     * The jobDefinition arn
     */
    private readonly jobDefinitionArn: string,

    /**
     * The JobQueue Resource
     */
    private readonly jobDefinitionScope: IConstruct,
    private readonly props: BatchJobProps = {},
  ) { }

  /**
   * Returns a RuleTarget that can be used to trigger queue this batch job as a
   * result from an EventBridge event.
   */
  public bind(rule: events.IRule, _id?: string): events.RuleTargetConfig {
    const batchParameters: events.CfnRule.BatchParametersProperty = {
      jobDefinition: this.jobDefinitionArn,
      jobName: this.props.jobName ?? Names.nodeUniqueId(rule.node),
      arrayProperties: this.props.size ? { size: this.props.size } : undefined,
      retryStrategy: this.props.attempts ? { attempts: this.props.attempts } : undefined,
    };

    if (this.props.deadLetterQueue) {
      addToDeadLetterQueueResourcePolicy(rule, this.props.deadLetterQueue);
    }

    return {
      ...bindBaseTargetConfig(this.props),
      arn: this.jobQueueArn,
      // When scoping resource-level access for job submission, you must provide both job queue and job definition resource types.
      // https://docs.aws.amazon.com/batch/latest/userguide/ExamplePolicies_BATCH.html#iam-example-restrict-job-def
      role: singletonEventRole(this.jobDefinitionScope, [
        new iam.PolicyStatement({
          actions: ['batch:SubmitJob'],
          resources: [
            this.jobDefinitionArn,
            this.jobQueueArn,
          ],
        }),
      ]),
      input: this.props.event,
      targetResource: this.jobQueueScope,
      batchParameters,
    };
  }
}
