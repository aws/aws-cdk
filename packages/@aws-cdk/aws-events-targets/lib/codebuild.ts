import * as codebuild from '@aws-cdk/aws-codebuild';
import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import { addToDeadLetterQueueResourcePolicy, bindBaseTargetConfig, singletonEventRole, TargetBaseProps } from './util';

/**
 * Customize the CodeBuild Event Target
 */
export interface CodeBuildProjectProps extends TargetBaseProps {

  /**
   * The role to assume before invoking the target
   * (i.e., the codebuild) when the given rule is triggered.
   *
   * @default - a new role will be created
   */
  readonly eventRole?: iam.IRole;

  /**
   * The event to send to CodeBuild
   *
   * This will be the payload for the StartBuild API.
   *
   * @default - the entire EventBridge event
   */
  readonly event?: events.RuleTargetInput;
}

/**
 * Start a CodeBuild build when an Amazon EventBridge rule is triggered.
 */
export class CodeBuildProject implements events.IRuleTarget {
  constructor(
    private readonly project: codebuild.IProject,
    private readonly props: CodeBuildProjectProps = {},
  ) {}

  /**
   * Allows using build projects as event rule targets.
   */
  public bind(_rule: events.IRule, _id?: string): events.RuleTargetConfig {

    if (this.props.deadLetterQueue) {
      addToDeadLetterQueueResourcePolicy(_rule, this.props.deadLetterQueue);
    }

    return {
      ...bindBaseTargetConfig(this.props),
      arn: this.project.projectArn,
      role: this.props.eventRole || singletonEventRole(this.project, [
        new iam.PolicyStatement({
          actions: ['codebuild:StartBuild'],
          resources: [this.project.projectArn],
        }),
      ]),
      input: this.props.event,
      targetResource: this.project,
    };
  }
}
