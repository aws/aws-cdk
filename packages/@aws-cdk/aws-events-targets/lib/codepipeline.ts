import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import { singletonEventRole } from './util';

/**
 * Customization options when creating a {@link CodePipeline} event target.
 */
export interface CodePipelineTargetOptions {
  /**
   * The role to assume before invoking the target
   * (i.e., the pipeline) when the given rule is triggered.
   *
   * @default - a new role will be created
   */
  readonly eventRole?: iam.IRole;
}

/**
 * Allows the pipeline to be used as an EventBridge rule target.
 */
export class CodePipeline implements events.IRuleTarget {
  constructor(
    private readonly pipeline: codepipeline.IPipeline,
    private readonly options: CodePipelineTargetOptions = {}) {
  }

  public bind(_rule: events.IRule, _id?: string): events.RuleTargetConfig {
    return {
      id: '',
      arn: this.pipeline.pipelineArn,
      role: this.options.eventRole || singletonEventRole(this.pipeline, [new iam.PolicyStatement({
        resources: [this.pipeline.pipelineArn],
        actions: ['codepipeline:StartPipelineExecution'],
      })]),
      targetResource: this.pipeline,
    };
  }
}
