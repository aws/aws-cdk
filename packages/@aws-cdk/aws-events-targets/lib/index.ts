export * from './batch';
export * from './codepipeline';
export * from './sns';
export * from './sqs';
export * from './codebuild';
export * from './aws-api';
export * from './lambda';
export * from './ecs-task-properties';
export * from './ecs-task';
export * from './state-machine';
