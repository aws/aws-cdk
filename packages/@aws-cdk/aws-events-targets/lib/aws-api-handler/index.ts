/* eslint-disable no-console */
// eslint-disable-next-line import/no-extraneous-dependencies
import * as AWS from 'aws-sdk';
import { AwsApiInput } from '../aws-api';

export async function handler(event: AwsApiInput) {
  console.log('Event: %j', event);
  console.log('AWS SDK VERSION: ' + (AWS as any).VERSION);

  const awsService = new (AWS as any)[event.service](event.apiVersion && { apiVersion: event.apiVersion });

  try {
    const response = await awsService[event.action](event.parameters).promise();
    console.log('Response: %j', response);
  } catch (e) {
    console.log(e);
    if (!event.catchErrorPattern || !new RegExp(event.catchErrorPattern).test(e.code)) {
      throw e;
    }
  }
}
