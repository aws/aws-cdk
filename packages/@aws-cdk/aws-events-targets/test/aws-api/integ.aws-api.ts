import * as events from '@aws-cdk/aws-events';
import * as cdk from '@aws-cdk/core';
import * as targets from '../../lib';

const app = new cdk.App();

class AwsApi extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Force new deployment of 'cool-service' and stop 'dev-instance' at midnight everyday
    const scheduleRule = new events.Rule(this, 'ScheduleRule', {
      schedule: events.Schedule.cron({
        hour: '0',
        minute: '0',
      }),
    });

    scheduleRule.addTarget(new targets.AwsApi({
      service: 'ECS',
      action: 'updateService',
      parameters: {
        service: 'cool-service',
        forceNewDeployment: true,
      } as AWS.ECS.UpdateServiceRequest,
    }));

    scheduleRule.addTarget(new targets.AwsApi({
      service: 'RDS',
      action: 'stopDBInstance',
      parameters: {
        DBInstanceIdentifier: 'dev-instance',
      } as AWS.RDS.StopDBInstanceMessage,
    }));

    // Create snapshots when a DB instance restarts
    const patternRule = new events.Rule(this, 'PatternRule', {
      eventPattern: {
        detailType: ['RDS DB Instance Event'],
        detail: {
          Message: ['DB instance restarted'],
        },
      },
    });

    patternRule.addTarget(new targets.AwsApi({
      service: 'RDS',
      action: 'createDBSnapshot',
      parameters: {
        DBInstanceIdentifier: events.EventField.fromPath('$.detail.SourceArn'),
      } as AWS.RDS.CreateDBSnapshotMessage,
    }));
  }
}

new AwsApi(app, 'aws-cdk-aws-api-target-integ');
app.synth();
