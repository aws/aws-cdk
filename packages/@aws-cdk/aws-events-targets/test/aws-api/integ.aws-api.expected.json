{
  "Resources": {
    "ScheduleRuleDA5BD877": {
      "Type": "AWS::Events::Rule",
      "Properties": {
        "ScheduleExpression": "cron(0 0 * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
                "Arn"
              ]
            },
            "Id": "Target0",
            "Input": "{\"service\":\"ECS\",\"action\":\"updateService\",\"parameters\":{\"service\":\"cool-service\",\"forceNewDeployment\":true}}"
          },
          {
            "Arn": {
              "Fn::GetAtt": [
                "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
                "Arn"
              ]
            },
            "Id": "Target1",
            "Input": "{\"service\":\"RDS\",\"action\":\"stopDBInstance\",\"parameters\":{\"DBInstanceIdentifier\":\"dev-instance\"}}"
          }
        ]
      }
    },
    "ScheduleRuleAllowEventRuleawscdkawsapitargetintegScheduleRule51140722763E20C1": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
            "Arn"
          ]
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "ScheduleRuleDA5BD877",
            "Arn"
          ]
        }
      }
    },
    "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRole9FFE9C50": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRoleDefaultPolicy4D43A7C1": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ecs:UpdateService",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": "rds:StopDBInstance",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": "rds:CreateDBSnapshot",
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRoleDefaultPolicy4D43A7C1",
        "Roles": [
          {
            "Ref": "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRole9FFE9C50"
          }
        ]
      }
    },
    "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters4e52413f31cff0a335f5083fa6197a6cb61928644842d89026c42c2d2a98342eS3Bucket38E36746"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters4e52413f31cff0a335f5083fa6197a6cb61928644842d89026c42c2d2a98342eS3VersionKeyFB07730C"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters4e52413f31cff0a335f5083fa6197a6cb61928644842d89026c42c2d2a98342eS3VersionKeyFB07730C"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRole9FFE9C50",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x"
      },
      "DependsOn": [
        "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRoleDefaultPolicy4D43A7C1",
        "AWSb4cf1abd4e4f4bc699441af7ccd9ec37ServiceRole9FFE9C50"
      ]
    },
    "PatternRule4AF6D328": {
      "Type": "AWS::Events::Rule",
      "Properties": {
        "EventPattern": {
          "detail-type": [
            "RDS DB Instance Event"
          ],
          "detail": {
            "Message": [
              "DB instance restarted"
            ]
          }
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
                "Arn"
              ]
            },
            "Id": "Target0",
            "InputTransformer": {
              "InputPathsMap": {
                "detail-SourceArn": "$.detail.SourceArn"
              },
              "InputTemplate": "{\"service\":\"RDS\",\"action\":\"createDBSnapshot\",\"parameters\":{\"DBInstanceIdentifier\":<detail-SourceArn>}}"
            }
          }
        ]
      }
    },
    "PatternRuleAllowEventRuleawscdkawsapitargetintegPatternRule3D388581AA4F776B": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
            "Arn"
          ]
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "PatternRule4AF6D328",
            "Arn"
          ]
        }
      }
    }
  },
  "Parameters": {
    "AssetParameters4e52413f31cff0a335f5083fa6197a6cb61928644842d89026c42c2d2a98342eS3Bucket38E36746": {
      "Type": "String",
      "Description": "S3 bucket for asset \"4e52413f31cff0a335f5083fa6197a6cb61928644842d89026c42c2d2a98342e\""
    },
    "AssetParameters4e52413f31cff0a335f5083fa6197a6cb61928644842d89026c42c2d2a98342eS3VersionKeyFB07730C": {
      "Type": "String",
      "Description": "S3 key for asset version \"4e52413f31cff0a335f5083fa6197a6cb61928644842d89026c42c2d2a98342e\""
    },
    "AssetParameters4e52413f31cff0a335f5083fa6197a6cb61928644842d89026c42c2d2a98342eArtifactHash3C551617": {
      "Type": "String",
      "Description": "Artifact hash for asset \"4e52413f31cff0a335f5083fa6197a6cb61928644842d89026c42c2d2a98342e\""
    }
  }
}